File: 4056-longest-balanced-substring-ii/longest-balanced-substring-ii.py
class Solution:
    def longestBalanced(self, s: str) -> int:
        n,ans=len(s),0
        i=0
        while i<n:
            j=i
            while j<n and s[j]==s[i]:
                j+=1
            ans=max(ans,j-i)
            i=j

        def best_2(x,y,third):
            best,i=0,0
            while i<n:
                if s[i]==third:
                    i+=1
                    continue
                st,bal=i,0
                fst={0:st}
                j=st
                while j<n and s[j]!=third:
                    bal+=1 if s[j]==x else -1
                    if bal not in fst:
                        fst[bal]=j+1
                    else:
                        best=max(best,j+1-fst[bal])
                    j+=1
                i=j
            return best

        ans=max(ans,best_2('a','b','c'))
        ans=max(ans,best_2('a','c','b'))
        ans=max(ans,best_2('b','c','a'))

        ca,cb,cc=0,0,0
        mp={(0,0):-1}
        for i,ch in enumerate(s):
            if ch=='a':
                ca+=1
            elif ch=='b':
                cb+=1
            else:
                cc+=1
            k=(ca-cb,ca-cc)
            if k in mp:
                ans=max(ans,i-mp[k])
            else:
                mp[k]=i
        return ans
                

File: 4056-longest-balanced-substring-ii/longest-balanced-substring-ii.cpp
class Solution {
public:
    int n;

    int helper(string s, char x, char y) {
        int countx = 0, county = 0, maxL = 0;
        map<int,int> mp;
        mp[0] = -1;

        for (int i = 0; i < n; i++) {

            // break sequence if other character appears
            if (s[i] != x && s[i] != y) {
                mp.clear();
                countx = 0;
                county = 0;
                continue;
            }

            if (s[i] == x) countx++;
            else county++;

            if (countx == county) {
                maxL = max(maxL, countx + county);
                continue;
            }

            if (mp.find(countx - county) != mp.end()) {
                maxL = max(maxL, i - mp[countx - county]);
            } else {
                mp[countx - county] = i;
            }
        }

        return maxL;
    }

    int longestBalanced(string s) {
        n = s.size();
        int maxS = 0;

        // Case 1: single character longest run
        int count = 1;
        for (int i = 1; i < n; i++) {
            if (s[i] == s[i-1]) count++;
            else {
                maxS = max(maxS, count);
                count = 1;
            }
        }
        maxS = max(maxS, count);

        // Case 2: two-character balance
        maxS = max(maxS, helper(s, 'a', 'b'));
        maxS = max(maxS, helper(s, 'b', 'c'));
        maxS = max(maxS, helper(s, 'a', 'c'));

        // Case 3: all three characters balanced
        int countA = 0, countB = 0, countC = 0;
        map<pair<int,int>, int> diffMap;

        for (int i = 0; i < n; i++) {
            if (s[i] == 'a') countA++;
            if (s[i] == 'b') countB++;
            if (s[i] == 'c') countC++;

            if (countA == countB && countA == countC) {
                maxS = max(maxS, countA + countB + countC);
                continue;
            }

            int diffAB = countA - countB;
            int diffAC = countA - countC;
            pair<int,int> key = {diffAB, diffAC};

            if (diffMap.count(key)) {
                maxS = max(maxS, i - diffMap[key]);
            } else {
                diffMap[key] = i;
            }
        }

        return maxS;
    }
};